### ISSUE TITLE
Bug: [no-unnecessary-type-arguments]: False positive on function/constructor with default generic parameter passed as value

### ISSUE THREAD

**Issue by 6XGate** [typescript-eslint/typescript-eslint#7042]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqAfFALxQIQDuUAFAHSMCGqA5kgFxRMIgDaAXQCUZEgQBQ4gGboE2AJaIoANwip5UkAEkkhKBAAewCAgAmSKHABGAKwjYitbLi7xkaLDnzERAb3FQUKgQwOioCHTKTAA26BBcsgDWCHBUCEJcUbHQ8hYEoioxcVDy7jyYEHBSUM6o4gC+kpjRTEgWAEoQUnrkKOoIrCT+gVlxXBKBmIh9nri0o-FQBH4BgVDAABa59AtkhdmrjY3iU+5BUhbklDSdUrQARMYo9yKtUEkpaSfTwOcRV9QoLdaABWegAJleFg+qQQTR+UCsTFMe18JSQAGU0KVWLdMmoNNokLRbiJGho6EjTPRclj+riurRUBchCtAgB6dlQACiBkg2AgKLUqFwUA2aggABooEhShV9sVcpZqqBIFAAAY8EDq+irU5IODRCD0aJwVhMi47IrGuWxUwQYkAciewEdrIa8LOrDgcBR5DRtOxAzxKgJmh0JK6eD6OKIZPEFNo3t9NMxQYZd2ZSFZUGGNWmhuNpvNWat2RpcjtDtozodrvdx31v1yADl0ABbKxqW57VTqcPE254BAdruoOMJ6q0Vuj7uM5npNn55CFk1mi0IMtxeg4ABi8gMgtoDfEQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolp9oAc1gBbRC3RRE0aB2iQANOGx4ipZBWp0GzNp269BwvsmSUxTabIyQFSleuxQCJclRr1kjFnYuHn4hWjI+eHg5J0VlNQ0cTx09X0NAkxDzWgATRDJ4UT58Sg4mWmlxZjEY53i3TWTvfT8A42CzMJkpcI5JYlFKZDLauNdEj21mtP8jINNQ4WlJACMFETIyFFRHOvH3LS9dHwNZjI7F2mhEfFhoJlGXBIOm45b09tMmfO3RQlopiJxFIZPgdvIxpANABfEDQoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2RGwAqkXYl0AEwAysaAENUZSh269aA4aIkAFWZlnx0eLIraZOPGlHQAPWc1wFi8rZliIseAJ4AVT+5PKFnwCAJJIqEzIeABKhuyYpKyBqlACAGLyqHgAFq7sAObZAMKyYiQhRADKAO5R0NlJZiqW-EJE9ejQANYSFeI2EgCC0NDoJAFNQWqwYYgR0FHYAG5YjP2N5imtLm4ensOj42mu8H02lcj5RLJ48egbzXyZsNUaYlKyAEboCqwAZvIxFMoM9XkRMOg7NkvhgirB1pQAQp0LwAL4gVFAA

### Repro Code

```typescript
type Constructor<T> = new (...args: any[]) => T

function verifyIs<T extends object>(ctor: Constructor<T>) {
  return (value: unknown): value is T => value instanceof ctor
}

class Ref<T = string> {
  value: T
  constructor(value: T) {
    this.value = value
  }
}

const rfs = new Ref("test") as unknown
const rfn = new Ref(5.2) as unknown

const bad = { isStringRef: verifyIs(Ref) }
if (bad.isStringRef(rfs)) {
  // Expected error here, since value is of type `any`.
  console.log(rfs.value.includes('test'))
}

const good = { isStringRef: verifyIs(Ref<string>) }
if (good.isStringRef(rfs)) {
  console.log(rfs.value.includes('test'))
}

const isNumberRef = verifyIs(Ref<number>)
if (isNumberRef(rfn)) {
  console.log(rfn.value.toFixed())
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unsafe-argument": "error",
    "@typescript-eslint/no-unsafe-assignment": "error",
    "@typescript-eslint/no-unsafe-call": "error",
    "@typescript-eslint/no-unsafe-declaration-merging": "error",
    "@typescript-eslint/no-unsafe-enum-comparison": "error",
    "@typescript-eslint/no-unsafe-member-access": "error",
    "@typescript-eslint/no-unsafe-return": "error",
    "@typescript-eslint/no-unnecessary-type-arguments": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "exactOptionalPropertyTypes": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedIndexedAccess": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "allowUnusedLabels": false,
    "allowUnreachableCode": false
  }
}
```

### Expected Result

No error for the `good` `isStringRef`.

Due to a design limitation in TypeScript, passing a generic function or constructor to any variable or argument that has a default generic parameter will resolve said parameter to it's constrained types. For example, `T extends false | null = null` will resolve to `false | null` when stored in a variables or passed to a function. In the case of no constraints, the parameter will resolve to `any`.

To resolve this issue, the generic parameter MUST be specified, even if it is the default value.

### Actual Result

One line 23, a `no-unnecessary-type-arguments` is raised. If the type argument is removed, the result is like that of line 20, a `no-unsafe-call` with a `no-unsafe-member-access`.

**Comment by bradzacher** [comment]
Huh, that's really interesting behaviour.

Whem referenced solely as a type, then TS uses the default value for the generic. However when referenced as an expression, then TS uses `any` for the generic.

So for an "expression with type arguments" we should probably just not apply the rule as it seems its always required?

**Comment by 6XGate** [comment]
That seems to be case. And I could swear I found a ticket for this in TypeScript that was closed as **Working as Intended** or **Design Limitation**, but can't find it anymore.

I thought I had some code in my company's repo or a personal project that landed on this TypeScript behavior and marked it with a `HACK` or `NOTE` comment with a link to the ticket, but alas.

**Comment by axelboc** [comment]
I think I'm experiencing this issue as well. Here is a contrived example to start with:

```ts
function myFunc<T extends string = string>(value: T): T {
  return value;
}

const value = myFunc(''); // type of value is inferred as ''

// But I want `value` to have type `string`, so two options:

// 1. Pass a parameter of type `string`
const value = myFunc('' as string);

// 2. Specify the generic explicitly (even though `string` is already the default)
const value = myFunc<string>('');
```

With the second option, `no-unnecessary-type-arguments` reports that `<string>` is unnecessary.

---

The use case I'm dealing with is a `useParam` React hook that allows controlling a URL query parameter. It returns the current value of the query param and a setter to change it. It also accepts a `defaultValue` as second parameter, which is where the generic is used. In some cases, the query param can receive any string value, but in other cases, I want to restrict it to specific strings (e.g. `'list' | 'grid'`):

```ts
const [filters, setFilters] = useParam<string>('filters', '');
const [view, setView] = useParam<'list' | 'summary'>('view', 'list');

function useParam<T extends string = string>(paramName: string, defaultValue: T): [T, (value: T) => void] {  ... }
```

If I remove `<string>` as this linting rule suggests, `filters` is inferred as `''`, which breaks the rest of the code.
